x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,140 @@
+
+    public static final CoordinatorProxyAction INSTANCE = new CoordinatorProxyAction();
+    public static final String NAME = "indices:data/read/xpack/enrich/coordinate_lookups";
+
+    private CoordinatorProxyAction() {
+        super(NAME, SearchResponse::new);
+    }
+
+    public static class TransportAction extends HandledTransportAction<SearchRequest, SearchResponse> {
+
+        private final Coordinator coordinator;
+
+        @Inject
+        public TransportAction(TransportService transportService, ActionFilters actionFilters, Coordinator coordinator) {
+            super(NAME, transportService, actionFilters, (Writeable.Reader<SearchRequest>) SearchRequest::new);
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> listener) {
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
+
+        final Client client;
+        final int maxLookupsPerRequest;
+        final int maxNumberOfConcurrentRequests;
+        final BlockingQueue<Item> queue = new LinkedBlockingQueue<>();
+
+        int numberOfOutstandingRequests = 0;
+
+        public Coordinator(Client client, Settings settings) {
+            this(client,
+                EnrichPlugin.COORDINATOR_PROXY_MAX_LOOKUPS_PER_REQUEST.get(settings),
+                EnrichPlugin.COORDINATOR_PROXY_MAX_CONCURRENT_REQUESTS.get(settings));
+        }
+
+        Coordinator(Client client, int maxLookupsPerRequest, int maxNumberOfConcurrentRequests) {
+            this.client = client;
+            this.maxLookupsPerRequest = maxLookupsPerRequest;
+            this.maxNumberOfConcurrentRequests = maxNumberOfConcurrentRequests;
+        }
+
+        void schedule(SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
+            queue.add(new Item(searchRequest, listener));
+            coordinateLookups();
+        }
+
+        synchronized void coordinateLookups() {
+            while (queue.isEmpty() == false &&
+                numberOfOutstandingRequests <= maxNumberOfConcurrentRequests) {
+
+                final List<Item> items = new LinkedList<>();
+                queue.drainTo(items, maxLookupsPerRequest);
+                final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+                items.forEach(item -> multiSearchRequest.add(item.searchRequest));
+
+                numberOfOutstandingRequests++;
+                client.multiSearch(multiSearchRequest, ActionListener.wrap(response -> {
+                    handleResponse(items, response, null);
+                }, e -> {
+                    handleResponse(items, null, e);
+                }));
+            }
+        }
+
+        synchronized void handleResponse(List<Item> items, MultiSearchResponse response, Exception e) {
+            numberOfOutstandingRequests--;
```

Comment:
I think this is the only line from this method call that needs to be protected by the lock? If this is true then maybe an AtomicInteger would help never having to wait on a lock when handling responses?